### PR TITLE
Support integer literals of different radices and decimal numeric literals

### PR DESCRIPTION
**Pull request by chengluyu**
This PR adds the support for various literals.

- Hexadecimal integer literals: `0xCAFE`, etc.
- Octal integer literals: `0o4567`, etc.
- Binary integer literals: `0b0010`, etc.
- Decimal numeric literals in floating-point style: `1e100`, `3.141592653589793`, etc.
- Support separators to be interleaved in _all_ digits: `0x0010_1101`, `12_34.56_78E+9_0`, etc.
- Lastly, the support for string escapes `\\`, `\b`, and `\f`. (I forgot this so it's not indicated by my commit messages. 🤦‍♂️)

> [!NOTE]  
> This PR does not support numeric literals starting with decimal points, for example, `.5`, because it conflicts with selecting tuple indices. We can check the accumulated tokens to determine the meaning of tokens such as `.1`, but I don't have time to summarize various situations and add tests right now, so I will leave it for future work.

### CODE REVIEW

**Comment by chengluyu** [comment]
[6182f20](https://github.com/hkust-taco/mlscript/pull/199/commits/6182f207255d026a2260c509bf8b050ac21d329b) failed for some reasons. Strange.

**Comment by NeilKleistGao** [comment]
> [6182f20](https://github.com/hkust-taco/mlscript/pull/199/commits/6182f207255d026a2260c509bf8b050ac21d329b) failed for some reasons. Strange.

Random-happened time-out problem lol.

File: shared/src/main/scala/mlscript/NewLexer.scala [hunk 1]
```diff
@@ -59,15 +65,113 @@ class NewLexer(origin: Origin, raise: Diagnostic => Unit, dbg: Bool) {
...
+      val (fraction, j) = if (i < length && bytes(i) === '.') {
+        takeDigits(i + 1, isDigit) match {
+          case (N, j) =>
+            raise(ErrorReport(msg"Expect at least one digit after the decimal point" -> S(loc(i + 1, i + 2)) :: Nil,
+              newDefs = true, source = Lexing))
+            ("", j)
+          case (S(digits), j) => ("." + digits, j)
+        }
+      } else ("", i)
+      val (exponent, k) = if (j < length && (bytes(j) === 'e' || bytes(j) === 'E')) {
```
Comment (NeilKleistGao) [inline shared/src/main/scala/mlscript/NewLexer.scala]: Maybe a helper function for `i < length && bytes(i) === something`? :thinking:

File: shared/src/main/scala/mlscript/NewLexer.scala [hunk 2]
```diff
@@ -59,15 +65,113 @@ class NewLexer(origin: Origin, raise: Diagnostic => Unit, dbg: Bool) {
...
+      bytes(i) match {
+        case '0' if i + 1 < length => bytes(i + 1) match {
+          case 'x' => integer(i + 2, 16, "hexadecimal", isHexDigit)
+          case 'o' => integer(i + 2, 8, "octal", isOctDigit)
+          case 'b' => integer(i + 2, 2, "binary", isBinDigit)
+          case '.' | 'E' | 'e' => decimal(i + 1, "0")
+          case _ => integer(i, 10, "decimal", isDigit)
+        }
+        case '0' => (zero, i + 1)
+        // case '.' => decimal(i, "0")
```
Comment (LPTK) [inline shared/src/main/scala/mlscript/NewLexer.scala]: Remove comment?

File: shared/src/main/scala/mlscript/NewLexer.scala [hunk 3]
```diff
@@ -59,15 +65,113 @@ class NewLexer(origin: Origin, raise: Diagnostic => Unit, dbg: Bool) {
...
+          else (acc.reverseIterator.mkString, firstSep, lastSep, i)
+        }
+        else (acc.reverseIterator.mkString, firstSep, lastSep, i)
+      val (str, firstSep, lastSep, j) = rec(i, Nil, false, false)
+      if (firstSep)
+        raise(WarningReport(
+          msg"Leading separator is not allowed" -> S(loc(i - 1, i)) :: Nil,
+          newDefs = true,
+          source = Lexing
+        ))
```
Comment (LPTK) [inline shared/src/main/scala/mlscript/NewLexer.scala]:
```suggestion
          newDefs = true, source = Lexing))
```

File: shared/src/main/scala/mlscript/NewLexer.scala [hunk 4]
```diff
@@ -59,15 +65,113 @@ class NewLexer(origin: Origin, raise: Diagnostic => Unit, dbg: Bool) {
...
+      if (lastSep)
+        raise(WarningReport(
+          msg"Trailing separator is not allowed" -> S(loc(j - 1, j)) :: Nil,
+          newDefs = true,
+          source = Lexing
+        ))
```
Comment (LPTK) [inline shared/src/main/scala/mlscript/NewLexer.scala]:
```suggestion
          newDefs = true, source = Lexing))
```

**Review by LPTK — APPROVED** [review]
LGTM besides @NeilKleistGao's comment.

**Pull request merged**